src/Microsoft.AspNet.Mvc.Razor/Precompilation/PrecompilationTagHelperTypeResolver.cs

```diff
@@ -0,0 +1,138 @@
+            {
+                foreach (var reference in _compilation.References)
+                {
+                    var compilationReference = reference as CompilationReference;
+                    if (compilationReference != null &&
+                        string.Equals(
+                            compilationReference.Compilation.AssemblyName,
+                            assemblyName,
+                            StringComparison.Ordinal))
+                    {
+                        return GetExportedTypes(compilationReference.Compilation.Assembly);
+                    }
+
+                    var assemblyReference = reference as PortableExecutableReference;
+                    if (assemblyReference != null)
+                    {
+                        var assemblyOrModuleSymbol = _compilation.GetAssemblyOrModuleSymbol(reference);
+                        if (assemblyOrModuleSymbol != null &&
+                            assemblyOrModuleSymbol.Kind == SymbolKind.Assembly)
+                        {
+                            var assemblySymbol = (IAssemblySymbol)assemblyOrModuleSymbol;
+                            if (string.Equals(
+                                    assemblySymbol.Identity.Name,
+                                    assemblyName,
+                                    StringComparison.Ordinal))
+                            {
+                                return GetExportedTypes(assemblySymbol);
+                            }
+                        }
+                    }
+                }
+
+                return Enumerable.Empty<ITypeInfo>();
+            }
+        }
+
+        private List<ITypeInfo> GetExportedTypes(IAssemblySymbol assembly)
+        {
+            var exportedTypes = new List<ITypeInfo>();
+            GetExportedTypes(assembly.GlobalNamespace, exportedTypes);
+            return exportedTypes;
+        }
+
+        private void GetExportedTypes(INamespaceSymbol namespaceSymbol, List<ITypeInfo> exportedTypes)
+        {
+            foreach (var type in namespaceSymbol.GetTypeMembers().Where(IsPublicType))
+            {
+                exportedTypes.Add(new SymbolBasedTypeInfo(type, OpenGenericDictionaryType));
+
+                // TODO: Nested types get eventually filtered out. Could we make it part of the contract?
+                var nestedPublicTypes = type.GetTypeMembers()
+                    .Where(IsPublicType)
+                    .Select(nestedType => new SymbolBasedTypeInfo(nestedType, OpenGenericDictionaryType));
+
+                exportedTypes.AddRange(nestedPublicTypes);
+            }
+
+            foreach (var subNamespace in namespaceSymbol.GetNamespaceMembers())
+            {
+                GetExportedTypes(subNamespace, exportedTypes);
+            }
+        }
+
+        private static bool IsPublicType(INamedTypeSymbol type) =>
+            type.TypeKind == TypeKind.Class &&
+            type.DeclaredAccessibility == Accessibility.Public;
```

Comment:
This doesn't seem to have the same restrictions as the Razor `TagHelperTypeResolver` unless Roslyn's `DeclaredAccessibility` is super overloaded:

https://github.com/aspnet/Razor/blob/dev/src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperTypeResolver.cs#L91-L95